File: Python/knowledge-base/knowledge_base/Embeddings/TextChunker.py
from typing import List


class TextChunker:
    """Character-based text chunker with configurable size and overlap."""

    def chunk_text(
            self,
            text: str,
            chunk_size: int = 500,
            overlap: int = 50) -> List[str]:
        """
        Split text into overlapping character-based chunks.

        Args:
            text: The input text to chunk.
            chunk_size: Maximum number of characters per chunk.
            overlap: Number of characters to overlap between consecutive chunks.

        Returns:
            List of text chunk strings.
        """
        if not text:
            return []

        if chunk_size <= 0:
            raise ValueError("chunk_size must be positive")

        if overlap < 0:
            raise ValueError("overlap must be non-negative")

        if overlap >= chunk_size:
            raise ValueError("overlap must be less than chunk_size")

        chunks = []
        start = 0
        text_length = len(text)

        while start < text_length:
            end = min(start + chunk_size, text_length)
            chunk = text[start:end].strip()
            if chunk:
                chunks.append(chunk)
            if end == text_length:
                break
            start = end - overlap

        return chunks

File: Rust/knowledge-base/src/ingestion/text_chunker.rs
/// Character-based text chunker with configurable size and overlap.
///
/// Mirrors the Python `TextChunker.chunk_text` logic exactly:
/// - slides a window of `chunk_size` characters,
/// - steps forward by `chunk_size - overlap` each iteration,
/// - strips whitespace from each chunk,
/// - skips empty chunks.
#[derive(Debug, Clone)]
pub struct TextChunker {
    pub chunk_size: usize,
    pub overlap: usize,
}

impl TextChunker {
    /// Create a new `TextChunker`.
    ///
    /// # Panics
    /// Panics if `chunk_size == 0` or `overlap >= chunk_size`.
    pub fn new(chunk_size: usize, overlap: usize) -> Self {
        assert!(chunk_size > 0, "chunk_size must be positive");
        assert!(overlap < chunk_size, "overlap must be less than chunk_size");
        Self { chunk_size, overlap }
    }

    /// Split `text` into overlapping character-based chunks.
    ///
    /// Returns an empty `Vec` if `text` is empty.
    pub fn chunk_text(&self, text: &str) -> Vec<String> {
        if text.is_empty() {
            return Vec::new();
        }

        // Work over char indices to handle multibyte UTF-8 correctly.
        let chars: Vec<char> = text.chars().collect();
        let text_len = chars.len();
        let step = self.chunk_size - self.overlap;

        let mut chunks = Vec::new();
        let mut start = 0usize;

        while start < text_len {
            let end = (start + self.chunk_size).min(text_len);
            let chunk: String = chars[start..end].iter().collect();
            let chunk = chunk.trim().to_string();
            if !chunk.is_empty() {
                chunks.push(chunk);
            }
            if end == text_len {
                break;
            }
            start += step;
        }

        chunks
    }
}

impl Default for TextChunker {
    fn default() -> Self {
        Self::new(500, 50)
    }
}
